### PR TITLE
Extend automatic type coercion support in Iceberg table creation to char

### PR DESCRIPTION
**Pull request by SemionPar**
## Description
- extend automatic type coercion support in Iceberg table creation to `char` type
- refactor existing tests to reduce duplication (these were essentially the same)

## Additional context and related issues
https://github.com/trinodb/trino/pull/13981
https://github.com/trinodb/trino/pull/20611

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Automatically use `varchar` as type during table creation when 
  `char` is specified. ({issue}`19336`). ({issue}`21515`)
```

### CODE REVIEW

**Comment by findinpath** [comment]
https://github.com/trinodb/trino/actions/runs/8646870078/job/23709683367?pr=21515
Failure sounds related to your changes:
```
TestIcebergParquetConnectorTest>BaseIcebergConnectorTest.testCharVarcharComparison:304 
Expecting message to be:
  "Type not supported for Iceberg: char(3)"
but was:
  "For query 20240411_142226_03881_g6k7z: 
 SELECT k, v FROM test_char_varcharjwzwbc93wu WHERE v = CAST('  ' AS varchar(2))
not equal
Actual rows (up to 100 of 0 extra rows shown, 0 rows in total):
    
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [3,    ]
"
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -931,6 +933,9 @@ public Optional<io.trino.spi.type.Type> getSupportedType(ConnectorSession sessio
         if (type instanceof TimeType) {
             return Optional.of(TIME_MICROS);
         }
+        if (type instanceof CharType) {
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:936]: Right. On the other hands, Iceberg connector always uses unbounded varchar. 

https://github.com/trinodb/trino/blob/b0b49e0c4bdc0c0903483c4afa11e262d17991e9/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TypeConverter.java#L84-L85

https://github.com/trinodb/trino/blob/b0b49e0c4bdc0c0903483c4afa11e262d17991e9/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TypeConverter.java#L135-L137

[the remark this answers is not in the document]
Comment (SemionPar) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:936]: Changing to unbounded varchar!

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 2]
```diff
@@ -5643,7 +5643,7 @@ private List<DataMappingTestSetup> testDataMappingSmokeTestData()
                 .add(new DataMappingTestSetup("timestamp(3) with time zone", "TIMESTAMP '2020-02-12 15:03:00 +01:00'", "TIMESTAMP '9999-12-31 23:59:59.999 +12:00'"))
                 .add(new DataMappingTestSetup("timestamp(6) with time zone", "TIMESTAMP '1969-12-31 15:03:00.123456 +01:00'", "TIMESTAMP '1969-12-31 17:03:00.123456 +01:00'"))
                 .add(new DataMappingTestSetup("timestamp(6) with time zone", "TIMESTAMP '2020-02-12 15:03:00 +01:00'", "TIMESTAMP '9999-12-31 23:59:59.999999 +12:00'"))
-                .add(new DataMappingTestSetup("char(3)", "'ab'", "'zzz'"))
+                .add(new DataMappingTestSetup("char(3)", "'ab '", "'zzz'"))
```
Comment (SemionPar) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:5646]: `testDataMappingSmokeTest.DataMappingTestSetup("char(3)", "'ab'", "'zzz'")` is failing due to the fact that char is padded upon coercion:

```sql
CREATE TABLE test_data_mapping_smoke_char_3_63hssk8sp8 
    AS SELECT CAST(row_id AS varchar(50)) row_id, CAST(value AS char(3)) value, CAST(value AS char(3)) another_column 
    FROM (VALUES   ('null value', NULL),   ('sample value', 'ab'),   ('high value', 'zzz'))  t(row_id, value);

DESCRIBE test_data_mapping_smoke_char_3_63hssk8sp8;
-- MaterializedResult{rows=[[row_id, varchar, , ], [value, varchar, , ], [another_column, varchar, , ]]}

SELECT value FROM test_data_mapping_smoke_char_3_63hssk8sp8;
-- value='ab '

-- from assertion:
SELECT row_id FROM test_data_mapping_smoke_char_3_63hssk8sp8 WHERE rand() = 42 OR value = 'ab'
-- and expect VALUES 'sample value' -> failure as 'ab' != 'ab '
```

Changing it here breaks it for other connectors:

```
Error:    TestClickHouseConnectorTest>BaseConnectorTest.testDataMappingSmokeTest:5556->BaseConnectorTest.testDataMapping:5590->AbstractTestQueryFramework.assertQuery:350 For query 20240415_123625_02027_285wm: 
 SELECT row_id FROM test_data_mapping_smoke_char_3_mhz2ny6ak4 WHERE rand() = 42 OR value = 'ab '
not equal
Actual rows (up to 100 of 0 extra rows shown, 0 rows in total):
    
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [sample value]
```

I will use `filterDataMappingSmokeTestData` and change 'ab' -> 'ab ' just for Iceberg tests.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 3]
```diff
@@ -300,8 +300,23 @@ public void testDeleteOnV1Table()
...
+                getQueryRunner()::execute,
+                "test_char_varchar",
+                "(k, v) AS VALUES" +
+                        "   (-1, CAST(NULL AS CHAR(3))), " +
+                        "   (3, CAST('   ' AS CHAR(3)))," +
+                        "   (6, CAST('x  ' AS CHAR(3)))")) {
+            // varchar of length shorter than column's length
+            assertThat(query("SELECT k, v FROM " + table.getName() + " WHERE v = CAST('  ' AS varchar(2))")).returnsEmptyResult();
+            // varchar of length longer than column's length
+            assertThat(query("SELECT k, v FROM " + table.getName() + " WHERE v = CAST('  ' AS varchar(4))")).returnsEmptyResult();
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:314]: 4 blanks instead of 2

**Review by findinpath — APPROVED** [review]

**Review by pajaks — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 4]
```diff
@@ -931,6 +933,9 @@ public Optional<io.trino.spi.type.Type> getSupportedType(ConnectorSession sessio
...
+            return Optional.of(VARCHAR);
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:937]: SGTM

**Review by ebyhr — APPROVED** [review]

**Pull request merged**